specs/src/tx_client_v2.md

```diff
@@ -0,0 +1,148 @@
+Tx Client v2 enforces a per-account sequential submission queue that serializes signing, submission, and recovery, avoiding sequence races and minimizing transaction failures.
+
+## Protocol/Component Description
+
+### Per account sequential queue
+
+Every account gets its own **transaction queue**.
+
+This queue is in charge of everything for that account: taking new transactions(jobs), broadcasting them in submission order, reporting outcomes.
+
+Inside the queue we keep two kinds of transactions:
+
+- **Queued** - submitted but not broadcast yet
+- **Pending** - already broadcast and waiting for confirmation
+
+---
+
+### Broadcast and Confirmation loops
+
+Each queue runs two background loops:
+
+1. **A submitter**
+    - Picks the next queued tx, signs it, broadcasts it to the node.
+    - Listens for evicted transactions and resubmits them.
+2. **A monitor**
+    Checks the statuses of pending txs and knows how to manage evictions/rejections/confirmations.
+
+---
+
+### Submission flow
+
+1. Accepting a transaction(job)
+
+When a user submits a transaction (e.g. via `SubmitPayForBlobToQueue`):
+
+- Tx is wrapped in a job containing:
+  - transaction data
+  - user options
+  - context
+  - a result channel
+- The queue enforces memory based backpressure:
+  - If sufficient memory is available, the job is appended to the queue
+  - If queue exceeds its memory limit, submission blocks until memory is freed up
+ User should be able to set desired queue size during client initialization.
+- Nothing gets signed yet.
+- No network operations are performed.
+- User will receive tx result asynchronously via the result channel.
+
+2. Selecting the next transaction to broadcast
+
+The submitter processes transactions strictly in submission order:
+
+- Submission proceeds only if the queue is not in recovery mode.
+- The next unbroadcast transaction is selected.
+- The transaction is built and signed using the next sequence number.
+- The transaction is broadcast to the node.
+
+If broadcast succeeds:
+
+- Tx hash, sequence, and signed bytes are recorded directly on the queued tx.
+- Tx is now considered pending.
+
+If broadcast fails due to a sequence mismatch:
+
+- Submitter pauses further submissions.
+- Job remains in the queue.
+- Waits for confirmation/recovery logic to unblock submissions.
+
+Sequence mismatches during broadcast are treated as coordination signals not terminal errors. If client remains blocked for longer than a minute, then it's possible to parse expected sequence from the error message and resign as recovery mech.
+
+If broadcast fails for any other reason:
+
+- Tx is removed from the queue.
+- An error is returned to the user via the result channel.
+
+### Phase 2: Confirmation
+
+Confirmation runs asynchronously at a configurable polling interval.
+
+The monitor periodically queries the chain for the status of broadcast transactions, performing status checks in bounded batches starting from the first pending transaction.
+
```

Comment:
it might be worth clarifying if v2 expects to get all of the updates at once (via the new TxStatus rpc method).

this can dramatically simply the logic in that we can be confident that we know which items in the queue need to be resubmitted

Reply:
I will add.

Reply:
https://github.com/celestiaorg/celestia-app/pull/6313/commits/a6b61dd6eea295af8833ce8e6cdbb58ff2d857ac